server/lib/two-factor-authentication/webauthn.ts

```diff
@@ -0,0 +1,278 @@
+  });
+
+  if (!verifyResponse.verified) {
+    throw new Error('Invalid registration result.');
+  }
+
+  const metadata = await getAuthenticatorMetadata(verifyResponse);
+
+  for (const report of metadata?.statusReports ?? []) {
+    if (report.status === 'REVOKED') {
+      throw new Error('Authenticator not supported.');
+    }
+  }
+
+  return verifyResponse;
+}
+
+export async function authenticationOptions(user: User, req) {
+  const methods = await UserTwoFactorMethod.findAll<UserTwoFactorMethod<TwoFactorMethod.WEBAUTHN>>({
+    where: {
+      UserId: user.id,
+      method: TwoFactorMethod.WEBAUTHN,
+    },
+  });
+
+  const allowCredentials = methods.map(
+    method =>
+      <PublicKeyCredentialDescriptorFuture>{
+        id: Buffer.from(method.data.credentialId, 'base64url'),
+        type: 'public-key',
+      },
+  );
+
+  const options = simplewebauthn.generateAuthenticationOptions({
+    allowCredentials,
+    rpID: config.webauthn.rpId,
+    userVerification: 'discouraged',
+    timeout: WebAuthnTimeoutSeconds * 1000,
+  });
+
+  if (req.jwtPayload) {
+    await cache.set(
+      `webauth-authentication-challenge:${user.id}:${req.jwtPayload?.sessionId}`,
+      options.challenge,
+      WebAuthnTimeoutSeconds,
+    );
+  }
+
+  return options;
+}
+
+export async function verifyAuthenticationResponse(user: User, response: AuthenticationResponseJSON, req) {
+  const method = await UserTwoFactorMethod.findOne<UserTwoFactorMethod<TwoFactorMethod.WEBAUTHN>>({
+    where: {
+      UserId: user.id,
+      method: TwoFactorMethod.WEBAUTHN,
+      data: {
+        credentialId: response.id,
+      },
+    },
+  });
+
+  if (!method) {
+    throw new ApolloError('Two-factor authentication code is invalid', 'INVALID_2FA_CODE');
+  }
+
+  let expectedChallenge;
+  if (req.jwtPayload.scope === 'twofactorauth') {
+    expectedChallenge = req.jwtPayload?.authenticationOptions?.webauthn?.challenge;
+  } else {
+    expectedChallenge = await cache.get(`webauth-authentication-challenge:${user.id}:${req.jwtPayload?.sessionId}`);
+  }
+
+  const verificationResponse = await simplewebauthn.verifyAuthenticationResponse({
+    authenticator: {
+      counter: method.data.counter,
+      credentialID: Buffer.from(method.data.credentialId, 'base64url'),
+      credentialPublicKey: Buffer.from(method.data.credentialPublicKey, 'base64url'),
+    },
+    expectedChallenge,
+    expectedOrigin: config.webauthn.expectedOrigins,
+    expectedRPID: config.webauthn.rpId,
+    response,
+    advancedFIDOConfig: {
+      userVerification: 'discouraged',
+    },
+    requireUserVerification: false,
+  });
+
+  if (!verificationResponse.verified) {
+    throw new ApolloError('Two-factor authentication code is invalid', 'INVALID_2FA_CODE');
+  }
+
+  await method.update({
+    data: {
+      ...method.data,
+      counter: verificationResponse.authenticationInfo.newCounter,
+    },
+  });
+}
+
+async function getAuthenticatorMetadata(
```

Comment:
This whole function is cryptic, could we have comments explaining what it's doing and implementing?